compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -425,7 +465,11 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                         | hir::ItemKind::Const(_, generics, _)
                         | hir::ItemKind::TraitAlias(generics, _),
                     ..
-                }) if !param_ty => {
+                }) if !param_ty
+                    && (generics.params.iter().any(|param| {
+                        v.params.iter().any(|p| p.name == param.name.ident().name)
+                    }) || v.params.is_empty()) =>
+                {
```

Comment:
Don't you want this?
```suggestion
                }) if !param_ty
                    && v.params.iter().all(|param| {
                        generics.params.iter().any(|p| p.name == param.name.ident().name)
                    }) =>
                {
```

Reply:
I'm confused by the logic: what's a suggestion you're trying to avoid emitting by adding this condition?

Reply:
We currently suggest

```
help: consider introducing a `where` clause, but there might be an alternative better way to express this requirement
   |
15 | impl B where B: From<T> {
   |        ++++++++++++++++
```

Which is incorrect because the `impl` block doesn't have a type parameter `T`, but the *method* does:

```
help: consider further restricting the type
   |
LL |     pub fn from_many<T: Into<B> + Clone>(v: Vec<T>) -> Self where B: From<T> {
   |                                                             ++++++++++++++++
```

Reply:
With the logic as you've written it, it seems like the incorrect suggestion should be emitted because `generics.params.iter().any(|param| {...})` will return true (the parameter `B` validates this predicate). What am I missing?